### PR TITLE
[USETUP] Speed up CONSOLE_SetStatusTextXV +17%

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
When watching ReactOS Setup, I found the status text appeared to be flickering.
JIRA issue: [CORE-18838](https://jira.reactos.org/browse/CORE-18838)
## Proposed changes

- Get more speed by omitting erasing and overwriting whole line.

```txt
(base/setup/usetup/consup.c:90) CONSOLE_SetStatusTextXV: 52333248
(base/setup/usetup/consup.c:103) CONSOLE_SetStatusTextXV_Improved: 43486080
```
+17% speed up in `CONSOLE_SetStatusTextXV`.

## TODO

- [x] Do tests.

## Screenshot

VirtualBox_ReactOS_18_02_2023_18_23_57

### CODE REVIEW

File: base/setup/usetup/consup.c [hunk 1]
```diff
@@ -459,6 +501,35 @@ CONSOLE_SetStatusTextXV(
...
+    IN SHORT x,
+    IN LPCSTR fmt,
+    IN va_list args)
+{
+    COORD coPos;
+    DWORD Written;
+    CHAR Buffer[128];
+    INT nLength;
+
+    memset(Buffer, ' ', xScreen);
```
Comment (HBelusca) [inline base/setup/usetup/consup.c:515]: Shouldn't there be more tests for buffer overflow?
Like for example,
`memset(Buffer, ' ', min(sizeof(Buffer), xScreen));` here,
and using `_vsnprintf` for the call below?

File: base/setup/usetup/consup.c [hunk 2]
```diff
@@ -63,6 +75,34 @@ CONSOLE_Init(VOID)
     xScreen = csbi.dwSize.X;
     yScreen = csbi.dwSize.Y;
 
+    /* TODO: Delete this test code later */
```
Comment (HBelusca) [inline base/setup/usetup/consup.c:78]: ok i think you can remove it now :)

**Review by HBelusca — COMMENTED** [review]
lgtm, I will give a final approval once code is cleaned.

File: base/setup/usetup/consup.c (hunk 2 above)
Comment (katahiromz) [reply base/setup/usetup/consup.c]: Removed.

File: base/setup/usetup/consup.c (hunk 1 above)
Comment (katahiromz) [reply base/setup/usetup/consup.c:438]: `_vsnprintf` is slower than `vsprintf`.

Comment (HBelusca) [reply base/setup/usetup/consup.c:438]: OK yeah since the other functions in this file don't really sanitize the lengths, let's go with that philosophy, for the time being...

File: base/setup/usetup/consup.c [hunk 3]
```diff
@@ -432,29 +432,23 @@ CONSOLE_SetStatusTextXV(
     COORD coPos;
     DWORD Written;
     CHAR Buffer[128];
+    INT nLength;
 
-    vsprintf(Buffer, fmt, args);
+    ASSERT(xScreen <= sizeof(Buffer));
+    memset(Buffer, ' ', xScreen);
+    nLength = vsprintf(&Buffer[x], fmt, args);
+    Buffer[x + nLength] = ' ';
```
Comment (HBelusca) [inline base/setup/usetup/consup.c:440]: Also, `ASSERT(x + nLength < sizeof(Buffer));`

File: base/setup/usetup/consup.c (hunk 1 above)
Comment (HBelusca) [reply base/setup/usetup/consup.c:438]: Using `min(sizeof(Buffer), xScreen)` here, and in the WriteConsoleOutputCharacterA call below, would be more correct, since it would allow a future hypothetical scenario where xScreen could be larger than what it is currently is.

File: base/setup/usetup/consup.c [hunk 4]
```diff
@@ -432,29 +432,23 @@ CONSOLE_SetStatusTextXV(
     COORD coPos;
     DWORD Written;
     CHAR Buffer[128];
+    INT nLength;
```
Comment (HBelusca) [inline base/setup/usetup/consup.c:435]: Move that before the coPos declaration.

**Review by HBelusca — CHANGES_REQUESTED** [review]

File: base/setup/usetup/consup.c (hunk 3 above)
Comment (katahiromz) [reply base/setup/usetup/consup.c:441]: Fixed.

File: base/setup/usetup/consup.c (hunk 4 above)
Comment (katahiromz) [reply base/setup/usetup/consup.c]: Fixed.

File: base/setup/usetup/consup.c [hunk 5]
```diff
@@ -429,32 +429,27 @@ CONSOLE_SetStatusTextXV(
...
-                                xScreen,
-                                coPos,
-                                &Written);
-
-    coPos.X = x;
-
     WriteConsoleOutputCharacterA(StdOutput,
                                  Buffer,
-                                 (ULONG)strlen(Buffer),
+                                 xScreen,
```
Comment (HBelusca) [inline base/setup/usetup/consup.c:452]: Possibly
```suggestion
                                 min(sizeof(Buffer), xScreen),
```

File: base/setup/usetup/consup.c [hunk 6]
```diff
@@ -429,32 +429,27 @@ CONSOLE_SetStatusTextXV(
     IN LPCSTR fmt,
     IN va_list args)
 {
+    INT nLength;
     COORD coPos;
     DWORD Written;
     CHAR Buffer[128];
 
-    vsprintf(Buffer, fmt, args);
+    ASSERT(xScreen <= sizeof(Buffer));
```
Comment (HBelusca) [inline base/setup/usetup/consup.c:437]: May now be useless, since `min(sizeof(Buffer), xScreen)` is used.

**Comment by katahiromz** [comment]
There is no flickering. LGTM.

**Comment by JoachimHenze** [comment]
https://build.reactos.org/#/builders/3/builds/20406 0.4.15-dev-5718-gf1de615 f1de615b9cdb2e267b97f7e55e8712bfba68bd26 VBox befo Katayamas commit [SYSREG] Stage took: 33.335868 seconds
https://build.reactos.org/#/builders/3/builds/20408 0.4.15-dev-5719-gacd3148 acd3148c1a52dfa1eb870d074d9071bbf2b67f26 VBox with Katayamas commit [SYSREG] Stage took: 33.146574 seconds

https://build.reactos.org/#/builders/6/builds/32937 0.4.15-dev-5718-gf1de615 f1de615b9cdb2e267b97f7e55e8712bfba68bd26 KVM  befo Katayamas commit [SYSREG] Stage took: 47.870212 seconds
https://build.reactos.org/#/builders/6/builds/32939 0.4.15-dev-5719-gacd3148 acd3148c1a52dfa1eb870d074d9071bbf2b67f26 KVM  with Katayamas commit [SYSREG] Stage took: 42.906221 seconds

**Comment by JoachimHenze** [comment]
Do you work with VBox or with KVM Katayama?

**Comment by katahiromz** [comment]
It's working.
